File: rust/crates/corlinman-hooks/src/event.rs
//! `HookEvent` — the wire-stable enum broadcast on the hook bus.
//!
//! Serialization uses `#[serde(tag = "kind")]` so JSON payloads carry an
//! explicit discriminant field and can evolve independently of Rust field
//! order. Consumers outside Rust (Python bridge, gateway admin UI) depend
//! on that stability.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum HookEvent {
    MessageReceived {
        channel: String,
        session_key: String,
        content: String,
        metadata: serde_json::Value,
        /// Phase 4 W2 B2 iter 8: canonical user_id resolved via the
        /// `IdentityStore` from `(channel, channel_user_id)`. `None`
        /// when the channel adapter / chat handler hasn't been wired
        /// to the resolver yet — downstream consumers (persona,
        /// evolution observer) treat absent as "fall back to legacy
        /// per-channel-key attribution".
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    MessageSent {
        channel: String,
        session_key: String,
        content: String,
        success: bool,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    MessageTranscribed {
        session_key: String,
        transcript: String,
        media_path: String,
        media_type: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    MessagePreprocessed {
        session_key: String,
        transcript: String,
        is_group: bool,
        group_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    SessionPatch {
        session_key: String,
        patch: serde_json::Value,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    AgentBootstrap {
        workspace_dir: String,
        session_key: String,
        files: Vec<String>,
    },
    GatewayStartup {
        version: String,
    },
    ConfigChanged {
        section: String,
        old: serde_json::Value,
        new: serde_json::Value,
    },
    /// Emitted once a remote/local tool invocation settles. `ok == false`
    /// implies `error_code` is populated with a short machine-readable
    /// identifier (e.g. `"timeout"`, `"disconnected"`, `"unsupported"`).
    ///
    /// Phase 4 W1.5 (next-tasks A1): `tenant_id` carries the source
    /// tenant when the chat / tool path knows it. `None` falls back
    /// to `default` when the EvolutionObserver persists the resulting
    /// signal — matching the schema column default. Wider tenant
    /// context propagation through the chat lifecycle is a follow-up;
    /// this field is the wire-shape commitment.
    ToolCalled {
        tool: String,
        runner_id: String,
        duration_ms: u64,
        ok: bool,
        error_code: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        tenant_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    /// An approval request was raised; admins may see it via SSE or the UI.
    ///
    /// `timeout_at_ms` is wall-clock milliseconds since the Unix epoch: the
    /// deadline after which the gate will auto-resolve to `timeout`.
    /// `args_preview` is truncated on the emitting side so the bus doesn't
    /// fan out large JSON payloads.
    ApprovalRequested {
        id: String,
        session_key: String,
        plugin: String,
        tool: String,
        args_preview: String,
        timeout_at_ms: u64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    /// Administrator (or auto-timeout) decided an approval.
    ///
    /// `decision` is one of `"allow"`, `"deny"`, or `"timeout"`. `decider`
    /// carries the operator identity for `allow`/`deny` and is `None` for
    /// `timeout`.
    ///
    /// Phase 4 W1.5 (next-tasks A1): `tenant_id` carries the source
    /// tenant when the approval gate knows it. See `ToolCalled` for
    /// the legacy default semantics.
    ApprovalDecided {
        id: String,
        decision: String,
        decider: Option<String>,
        decided_at_ms: u64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        tenant_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    /// Rate-limit rejected a request.
    ///
    /// `limit_type` identifies which dimension tripped (e.g. `"channel_qq"`,
    /// `"channel_telegram"`, `"tool"`). `retry_after_ms` is a coarse hint
    /// (0 when no clean retry window is known).
    RateLimitTriggered {
        session_key: String,
        limit_type: String,
        retry_after_ms: u64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        user_id: Option<String>,
    },
    /// A NodeBridge-connected device emitted a telemetry metric.
    ///
    /// `node_id` identifies the remote device, `metric` is the metric name
    /// (dot-separated path; no schema enforced at the bus layer), and
    /// `tags` carries the metric's dimensional labels. Kept `BTreeMap` so
    /// the serialized JSON key order is stable across emits — matters for
    /// golden-file based tests downstream.
    Telemetry {
        node_id: String,
        metric: String,
        value: f64,
        tags: BTreeMap<String, String>,
    },
    /// The `corlinman-scheduler` ran a `JobAction::Subprocess` job whose
    /// child exited 0 within the configured timeout. Phase 2 wave 2-B uses
    /// this for the `evolution_engine` daily job; the `EvolutionObserver`
    /// folds it into `evolution_signals` so a successful run shows up as
    /// input to the *next* run — closing the loop.
    EngineRunCompleted {
        run_id: String,
        proposals_generated: u64,
        duration_ms: u64,
    },
    /// The `corlinman-scheduler` ran a `JobAction::Subprocess` job that
    /// either exited non-zero or was killed by the runtime's timeout
    /// guard. `error_kind` is one of `"exit_code"`, `"timeout"`,
    /// `"spawn_failed"`. `exit_code` is `None` for spawn / timeout cases.
    EngineRunFailed {
        run_id: String,
        error_kind: String,
        exit_code: Option<i32>,
    },
    /// Phase 4 W4 D3 iter 9: a child subagent was admitted by the
    /// supervisor and is about to start running. Mirrors the shape of
    /// `EngineRunCompleted/Failed` so downstream observers can pattern-
    /// match the four `Subagent*` variants the same way.
    ///
    /// `parent_trace_id` is the child's link back to its parent's trace
    /// tree — the evolution observer's join query (design § "Open
    /// question 4") uses it to fold child signals into the same cluster
    /// as the parent's. The child also gets its own `trace_id` (which
    /// may be the same string when the parent didn't bump it; the
    /// supervisor never mints a new id, just passes through).
    SubagentSpawned {
        parent_session_key: String,
        child_session_key: String,
        child_agent_id: String,
        agent_card: String,
        depth: u8,
        parent_trace_id: String,
        tenant_id: String,
    },
    /// The child reasoning loop returned cleanly (any of `Stop`,
    /// `Length`, `Error` — `Timeout` and the pre-spawn rejections have
    /// dedicated variants for clarity).
    ///
    /// `finish_reason` carries the underlying `FinishReason` snake_case
    /// value verbatim so the operator UI / evolution adapter can branch
    /// without re-importing the enum.
    SubagentCompleted {
        parent_session_key: String,
        child_session_key: String,
        child_agent_id: String,
        finish_reason: String,
        elapsed_ms: u64,
        tool_calls_made: u32,
        parent_trace_id: String,
        tenant_id: String,
    },
    /// The child's wall-clock budget expired and the cooperative cancel
    /// path fired. Distinct variant (rather than a `Completed` with
    /// `finish_reason="timeout"`) so dashboards can red-flag timeouts
    /// without parsing the inner field. `elapsed_ms` is the *actual*
    /// time spent before the cancel — typically `≈ max_wall_seconds`.
    SubagentTimedOut {
        parent_session_key: String,
        child_session_key: String,
        child_agent_id: String,
        elapsed_ms: u64,
        parent_trace_id: String,
        tenant_id: String,
    },
    /// The supervisor refused the spawn at depth-cap check time. No
    /// `child_session_key` because the slot was never allocated; the
    /// `parent_session_key` + `attempted_depth` pair are enough to
    /// identify the would-be spawn.
    ///
    /// Concurrency / tenant-quota rejections share this variant with
    /// `reason` discriminating; matches the design's "all four caps emit
    /// hook events on rejection" wording (§ "Resource governance"). The
    /// `reason` strings are `"depth_capped"`, `"parent_concurrency_exceeded"`,
    /// `"tenant_quota_exceeded"` — the same lowercase shape the supervisor's
    /// own `AcquireReject::Display` produces.
    SubagentDepthCapped {
        parent_session_key: String,
        attempted_depth: u8,
        reason: String,
        parent_trace_id: String,
        tenant_id: String,
    },
}

impl HookEvent {
    /// Short, stable discriminant string. Used as a tracing span field so
    /// traces can be filtered by event kind without parsing the enum.
    pub fn kind(&self) -> &'static str {
        match self {
            Self::MessageReceived { .. } => "message_received",
            Self::MessageSent { .. } => "message_sent",
            Self::MessageTranscribed { .. } => "message_transcribed",
            Self::MessagePreprocessed { .. } => "message_preprocessed",
            Self::SessionPatch { .. } => "session_patch",
            Self::AgentBootstrap { .. } => "agent_bootstrap",
            Self::GatewayStartup { .. } => "gateway_startup",
            Self::ConfigChanged { .. } => "config_changed",
            Self::ToolCalled { .. } => "tool_called",
            Self::ApprovalRequested { .. } => "approval_requested",
            Self::ApprovalDecided { .. } => "approval_decided",
            Self::RateLimitTriggered { .. } => "rate_limit_triggered",
            Self::Telemetry { .. } => "telemetry",
            Self::EngineRunCompleted { .. } => "engine_run_completed",
            Self::EngineRunFailed { .. } => "engine_run_failed",
            Self::SubagentSpawned { .. } => "subagent_spawned",
            Self::SubagentCompleted { .. } => "subagent_completed",
            Self::SubagentTimedOut { .. } => "subagent_timed_out",
            Self::SubagentDepthCapped { .. } => "subagent_depth_capped",
        }
    }

    /// Session key if the event is scoped to one. `None` for global events
    /// (startup, config changes).
    pub fn session_key(&self) -> Option<&str> {
        match self {
            Self::MessageReceived { session_key, .. }
            | Self::MessageSent { session_key, .. }
            | Self::MessageTranscribed { session_key, .. }
            | Self::MessagePreprocessed { session_key, .. }
            | Self::SessionPatch { session_key, .. }
            | Self::AgentBootstrap { session_key, .. }
            | Self::ApprovalRequested { session_key, .. }
            | Self::RateLimitTriggered { session_key, .. } => Some(session_key.as_str()),
            // Subagent events surface the *child's* session_key (when
            // present) — the operator UI's tree visualisation keys on
            // it for the spawn-message expansion. Pre-spawn rejections
            // (DepthCapped) have no child session yet, so report the
            // parent's instead so the parent's row still highlights.
            Self::SubagentSpawned { child_session_key, .. }
            | Self::SubagentCompleted { child_session_key, .. }
            | Self::SubagentTimedOut { child_session_key, .. } => {
                Some(child_session_key.as_str())
            }
            Self::SubagentDepthCapped { parent_session_key, .. } => {
                Some(parent_session_key.as_str())
            }
            Self::GatewayStartup { .. }
            | Self::ConfigChanged { .. }
            | Self::ToolCalled { .. }
            | Self::ApprovalDecided { .. }
            | Self::Telemetry { .. }
            | Self::EngineRunCompleted { .. }
            | Self::EngineRunFailed { .. } => None,
        }
    }
}

File: python/packages/corlinman-hooks/src/corlinman_hooks/event.py
"""`HookEvent` — the wire-stable tagged-union broadcast on the hook bus.

Mirrors ``rust/crates/corlinman-hooks/src/event.rs``. The Rust crate uses
``#[serde(tag = "kind")]`` with PascalCase variant names so JSON payloads
carry an explicit discriminant field. We replicate that wire shape here
so the Python plane and the Rust gateway can speak the same JSON.

Each event variant is a nested dataclass under :class:`HookEvent`. The
``HookEvent`` umbrella class also serves as the static type for
"any hook event" — it is never instantiated directly; callers always
construct a concrete variant such as ``HookEvent.MessageReceived(...)``.

The ``kind()`` instance method returns the short snake_case discriminant
used for tracing / metric labels. The ``session_key()`` instance method
returns the scoped session key when one is present, or ``None`` for
global events.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field, fields
from typing import Any, ClassVar

__all__ = ["HookEvent"]


# ---------------------------------------------------------------------------
# Shared helpers (variant -> kind string + JSON dict round-trip).
# ---------------------------------------------------------------------------


def _none_to_default(value: Any, default: Any) -> Any:
    """Skip serializing ``None`` for optional fields (Rust uses
    ``skip_serializing_if = "Option::is_none"`` on those)."""
    return default if value is None else value


class _HookEventBase:
    """Common mixin for every concrete event variant.

    Subclasses set the class-level :attr:`KIND` to the snake_case
    discriminant string returned by :meth:`kind`. The class name itself
    is the PascalCase JSON ``kind`` value used in serialization (which
    mirrors Rust's default for ``#[serde(tag = "kind")]``).
    """

    KIND: ClassVar[str] = ""
    # Sequence of fields that should be omitted from the JSON output when
    # their value is ``None`` (mirrors ``skip_serializing_if =
    # "Option::is_none"``).
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ()

    def kind(self) -> str:
        return self.KIND

    def session_key(self) -> str | None:
        """Default: not session-scoped. Variants override as needed."""
        return None

    # PascalCase wire name. Each concrete variant overrides this with
    # the unprefixed class name (`_MessageReceived` -> `MessageReceived`)
    # so the serializer matches the Rust ``#[serde(tag = "kind")]``
    # default exactly. Set at class definition time below to avoid
    # repeating the string in every subclass body.
    VARIANT_NAME: ClassVar[str] = ""

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"kind": self.VARIANT_NAME}
        for f in fields(self):  # type: ignore[arg-type]
            value = getattr(self, f.name)
            if f.name in self.OPTIONAL_FIELDS and value is None:
                continue
            out[f.name] = value
        return out

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), separators=(",", ":"))


# ---------------------------------------------------------------------------
# Concrete event variants. Field names + types track the Rust source.
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _MessageReceived(_HookEventBase):
    channel: str
    session_key_: str
    content: str
    metadata: Any
    user_id: str | None = None

    KIND: ClassVar[str] = "message_received"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    # `session_key` is both a field name in Rust and the helper method
    # on `_HookEventBase`. We store the field as `session_key_` and
    # expose it via the method/property below to avoid the clash while
    # still serializing as the canonical `session_key` JSON key.
    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _MessageSent(_HookEventBase):
    channel: str
    session_key_: str
    content: str
    success: bool
    user_id: str | None = None

    KIND: ClassVar[str] = "message_sent"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _MessageTranscribed(_HookEventBase):
    session_key_: str
    transcript: str
    media_path: str
    media_type: str
    user_id: str | None = None

    KIND: ClassVar[str] = "message_transcribed"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _MessagePreprocessed(_HookEventBase):
    session_key_: str
    transcript: str
    is_group: bool
    group_id: str | None
    user_id: str | None = None

    KIND: ClassVar[str] = "message_preprocessed"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _SessionPatch(_HookEventBase):
    session_key_: str
    patch: Any
    user_id: str | None = None

    KIND: ClassVar[str] = "session_patch"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _AgentBootstrap(_HookEventBase):
    workspace_dir: str
    session_key_: str
    files: list[str] = field(default_factory=list)

    KIND: ClassVar[str] = "agent_bootstrap"

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _GatewayStartup(_HookEventBase):
    version: str

    KIND: ClassVar[str] = "gateway_startup"


@dataclass(frozen=True)
class _ConfigChanged(_HookEventBase):
    section: str
    old: Any
    new: Any

    KIND: ClassVar[str] = "config_changed"


@dataclass(frozen=True)
class _ToolCalled(_HookEventBase):
    tool: str
    runner_id: str
    duration_ms: int
    ok: bool
    error_code: str | None
    tenant_id: str | None = None
    user_id: str | None = None

    KIND: ClassVar[str] = "tool_called"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("tenant_id", "user_id")


@dataclass(frozen=True)
class _ApprovalRequested(_HookEventBase):
    id: str
    session_key_: str
    plugin: str
    tool: str
    args_preview: str
    timeout_at_ms: int
    user_id: str | None = None

    KIND: ClassVar[str] = "approval_requested"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _ApprovalDecided(_HookEventBase):
    id: str
    decision: str
    decider: str | None
    decided_at_ms: int
    tenant_id: str | None = None
    user_id: str | None = None

    KIND: ClassVar[str] = "approval_decided"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("tenant_id", "user_id")


@dataclass(frozen=True)
class _RateLimitTriggered(_HookEventBase):
    session_key_: str
    limit_type: str
    retry_after_ms: int
    user_id: str | None = None

    KIND: ClassVar[str] = "rate_limit_triggered"
    OPTIONAL_FIELDS: ClassVar[tuple[str, ...]] = ("user_id",)

    def session_key(self) -> str | None:
        return self.session_key_

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        d["session_key"] = d.pop("session_key_")
        return d


@dataclass(frozen=True)
class _Telemetry(_HookEventBase):
    node_id: str
    metric: str
    value: float
    tags: dict[str, str] = field(default_factory=dict)

    KIND: ClassVar[str] = "telemetry"

    def to_dict(self) -> dict[str, Any]:
        d = super().to_dict()
        # Rust uses BTreeMap so the serialized JSON key order is
        # lexicographic. Mirror that here by sorting the tag dict.
        d["tags"] = dict(sorted(self.tags.items()))
        return d


@dataclass(frozen=True)
class _EngineRunCompleted(_HookEventBase):
    run_id: str
    proposals_generated: int
    duration_ms: int

    KIND: ClassVar[str] = "engine_run_completed"


@dataclass(frozen=True)
class _EngineRunFailed(_HookEventBase):
    run_id: str
    error_kind: str
    exit_code: int | None

    KIND: ClassVar[str] = "engine_run_failed"


@dataclass(frozen=True)
class _SubagentSpawned(_HookEventBase):
    parent_session_key: str
    child_session_key: str
    child_agent_id: str
    agent_card: str
    depth: int
    parent_trace_id: str
    tenant_id: str

    KIND: ClassVar[str] = "subagent_spawned"

    def session_key(self) -> str | None:
        return self.child_session_key


@dataclass(frozen=True)
class _SubagentCompleted(_HookEventBase):
    parent_session_key: str
    child_session_key: str
    child_agent_id: str
    finish_reason: str
    elapsed_ms: int
    tool_calls_made: int
    parent_trace_id: str
    tenant_id: str

    KIND: ClassVar[str] = "subagent_completed"

    def session_key(self) -> str | None:
        return self.child_session_key


@dataclass(frozen=True)
class _SubagentTimedOut(_HookEventBase):
    parent_session_key: str
    child_session_key: str
    child_agent_id: str
    elapsed_ms: int
    parent_trace_id: str
    tenant_id: str

    KIND: ClassVar[str] = "subagent_timed_out"

    def session_key(self) -> str | None:
        return self.child_session_key


@dataclass(frozen=True)
class _SubagentDepthCapped(_HookEventBase):
    parent_session_key: str
    attempted_depth: int
    reason: str
    parent_trace_id: str
    tenant_id: str

    KIND: ClassVar[str] = "subagent_depth_capped"

    def session_key(self) -> str | None:
        return self.parent_session_key


# ---------------------------------------------------------------------------
# Public umbrella class: exposes variants as attributes mirroring Rust's
# ``HookEvent::Variant`` syntax in the source crate.
# ---------------------------------------------------------------------------


class HookEvent(_HookEventBase):
    """Tagged-union of every event broadcast on the bus.

    Concrete variants are exposed as nested attributes so call sites
    look like ``HookEvent.MessageReceived(...)`` to mirror the Rust
    ``HookEvent::MessageReceived { ... }`` constructor syntax.

    Use :meth:`from_dict` / :meth:`from_json` to rehydrate a wire
    payload back into the concrete variant.
    """

    MessageReceived = _MessageReceived
    MessageSent = _MessageSent
    MessageTranscribed = _MessageTranscribed
    MessagePreprocessed = _MessagePreprocessed
    SessionPatch = _SessionPatch
    AgentBootstrap = _AgentBootstrap
    GatewayStartup = _GatewayStartup
    ConfigChanged = _ConfigChanged
    ToolCalled = _ToolCalled
    ApprovalRequested = _ApprovalRequested
    ApprovalDecided = _ApprovalDecided
    RateLimitTriggered = _RateLimitTriggered
    Telemetry = _Telemetry
    EngineRunCompleted = _EngineRunCompleted
    EngineRunFailed = _EngineRunFailed
    SubagentSpawned = _SubagentSpawned
    SubagentCompleted = _SubagentCompleted
    SubagentTimedOut = _SubagentTimedOut
    SubagentDepthCapped = _SubagentDepthCapped

    # Registry of PascalCase variant name -> concrete dataclass. Keys
    # mirror the Rust ``#[serde(tag = "kind")]`` discriminant values
    # exactly. We build this once at module import so ``from_dict`` is
    # an O(1) dispatch.
    _VARIANTS: ClassVar[dict[str, type[_HookEventBase]]] = {
        "MessageReceived": _MessageReceived,
        "MessageSent": _MessageSent,
        "MessageTranscribed": _MessageTranscribed,
        "MessagePreprocessed": _MessagePreprocessed,
        "SessionPatch": _SessionPatch,
        "AgentBootstrap": _AgentBootstrap,
        "GatewayStartup": _GatewayStartup,
        "ConfigChanged": _ConfigChanged,
        "ToolCalled": _ToolCalled,
        "ApprovalRequested": _ApprovalRequested,
        "ApprovalDecided": _ApprovalDecided,
        "RateLimitTriggered": _RateLimitTriggered,
        "Telemetry": _Telemetry,
        "EngineRunCompleted": _EngineRunCompleted,
        "EngineRunFailed": _EngineRunFailed,
        "SubagentSpawned": _SubagentSpawned,
        "SubagentCompleted": _SubagentCompleted,
        "SubagentTimedOut": _SubagentTimedOut,
        "SubagentDepthCapped": _SubagentDepthCapped,
    }

    # Set of variants that store the JSON ``session_key`` field under
    # the dataclass attribute ``session_key_`` (because the umbrella
    # method shadows the name on the class). ``from_dict`` rewrites
    # the key for these variants when constructing.
    _SESSION_KEY_VARIANTS: ClassVar[frozenset[str]] = frozenset(
        {
            "MessageReceived",
            "MessageSent",
            "MessageTranscribed",
            "MessagePreprocessed",
            "SessionPatch",
            "AgentBootstrap",
            "ApprovalRequested",
            "RateLimitTriggered",
        }
    )

    def __init__(self) -> None:  # pragma: no cover — never instantiated
        raise TypeError(
            "HookEvent is a tagged-union umbrella; instantiate a concrete "
            "variant such as HookEvent.MessageReceived(...) instead."
        )

    @classmethod
    def _bind_variant_names(cls) -> None:
        """Populate :attr:`_HookEventBase.VARIANT_NAME` on each variant.

        Called once at module-import time from below. We can't set the
        attribute inline on the dataclass body without repeating the
        string, so we drive it from the registry instead.
        """
        for name, variant in cls._VARIANTS.items():
            variant.VARIANT_NAME = name

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> _HookEventBase:
        """Construct a variant from a wire-format JSON dict.

        ``payload['kind']`` must be one of the PascalCase variant names
        (matching the Rust serde tag).
        """
        kind = payload.get("kind")
        if not isinstance(kind, str):
            raise ValueError("missing or non-string 'kind' field")
        variant = cls._VARIANTS.get(kind)
        if variant is None:
            raise ValueError(f"unknown hook event kind: {kind!r}")
        kwargs = {k: v for k, v in payload.items() if k != "kind"}
        if kind in cls._SESSION_KEY_VARIANTS and "session_key" in kwargs:
            kwargs["session_key_"] = kwargs.pop("session_key")
        # Fill in optional fields that may have been skipped by the
        # serializer (Rust ``skip_serializing_if = "Option::is_none"``
        # path). The dataclass defaults to ``None`` for these, so we
        # only need to handle keys the caller did pass.
        return variant(**kwargs)

    @classmethod
    def from_json(cls, raw: str) -> _HookEventBase:
        return cls.from_dict(json.loads(raw))


# Bind each concrete variant's PascalCase wire name now that the registry
# is fully populated. Keeps the dataclass bodies above free of repeated
# string literals while still letting ``to_dict`` emit the correct
# ``kind`` discriminant for round-trips through ``from_dict``.
HookEvent._bind_variant_names()
